### PR TITLE
[REF] wrong-import-position, wrong-import-order: Skip try-import cases and nested imports

### PR DESCRIPTION
**Pull request by moylop260**
- [x] Fix https://github.com/PyCQA/pylint/issues/901
- [x] Add tests with nested cases for local, external and standard imports
- [x] Fix self-pylint files
- [x] Compare all python files vs flake8. 
  - Different output of pylint of this pr vs flake8 with pylint/test/functional/wrong_import_position2.py

<s> - [x] Reduce too-complex rate of `_check_imports_order` method </s>

- [x] Add all tests cases from https://github.com/PyCQA/pylint/issues/901#issuecomment-219348588

Flake8 output
```bash
flake8 --select=E402 pylint/test/functional/wrong_import_position.py
pylint/test/functional/wrong_import_position.py:22:1: E402 module level import not at top of file
pylint/test/functional/wrong_import_position.py:26:1: E402 module level import not at top of file
pylint/test/functional/wrong_import_position.py:32:1: E402 module level import not at top of file
pylint/test/functional/wrong_import_position.py:33:1: E402 module level import not at top of file
```

pylint output with this PR
```bash
python -m pylint -d all -e wrong-import-position --output-format=colorized --msg-template="{path}:{line}: [{msg_id}({symbol}), {obj}] {msg}" -r n pylint/test/functional/wrong_import_position.py
************* Module functional.wrong_import_position
pylint/test/functional/wrong_import_position.py:22: [C0413(wrong-import-position), ] Import "import six" should be placed at the top of the module
pylint/test/functional/wrong_import_position.py:26: [C0413(wrong-import-position), ] Import "import datetime" should be placed at the top of the module
pylint/test/functional/wrong_import_position.py:32: [C0413(wrong-import-position), ] Import "import scipy" should be placed at the top of the module
pylint/test/functional/wrong_import_position.py:33: [C0413(wrong-import-position), ] Import "import astroid" should be placed at the top of the module
```

Evidence image
screen shot 2016-05-15 at 12 17 16 pm

### CODE REVIEW

File: pylint/checkers/imports.py [hunk 1]
```diff
@@ -512,39 +514,44 @@ def _check_imports_order(self, node):
...
-            if modname.startswith('.'):
-                package = '.' + modname.split('.')[1]
-            else:
-                package = modname.split('.')[0]
-
-            import_category = isort_obj.place_module(package)
-            if import_category in ('FUTURE', 'STDLIB'):
-                std_imports.append((node, package))
-                wrong_import = extern_imports or local_imports
+            package = '.' + modname.split('.')[1] if modname.startswith('.') \
```
Comment (PCManticore) [inline pylint/checkers/imports.py]: Why did you change this? It was easier to understand as an if-else statement.

File: pylint/checkers/imports.py [hunk 2]
```diff
@@ -512,39 +514,44 @@ def _check_imports_order(self, node):
...
+                else modname.split('.')[0]
+            isort_category = isort_obj.place_module(package)
+            category = isort2keys.get(isort_category, '')
+            # to share the same pointer of memory in list.append sentences
```
Comment (PCManticore) [inline pylint/checkers/imports.py]: I don't understand this comment.

File: pylint/checkers/imports.py [hunk 3]
```diff
@@ -512,39 +514,44 @@ def _check_imports_order(self, node):
...
+            node_pack = (node, package)
+            imports.setdefault(category + '_all', []).append(node_pack)
```
Comment (PCManticore) [inline pylint/checkers/imports.py]: I think this code is overly complicated and I can't figure out what you're intending to do with it. First of all, you're appending _nested to the category, but I can't see where you're removing it, users shouldn't care about being _nested or _all. Second, I feel that the use of setdefault is superfluous, why don't you simply determine the imports using some variables instead of putting the nodes into a dictionary?

File: pylint/checkers/imports.py [hunk 4]
```diff
@@ -524,26 +534,30 @@ def _check_imports_order(self, node):
...
-                if wrong_import:
+                if wrong_import and not nested:
                     self.add_message('wrong-import-order', node=node,
                                      args=('standard import "%s"' % node.as_string(),
                                            '"%s"' % wrong_import[0][0].as_string()))
             elif import_category in ('FIRSTPARTY', 'THIRDPARTY'):
                 extern_imports.append((node, package))
-                wrong_import = local_imports
-                if wrong_import:
+                if not nested:
```
Comment (PCManticore) [inline pylint/checkers/imports.py]: This looks much better than the previous version, thank you.

Comment (moylop260) [inline pylint/checkers/imports.py]: Thank you for your feedback.
I'm checking all test cases vs flake8

**Comment by moylop260** [comment]
FYI
```bash
[root@0cd245d5b6d4]~/build/PyCQA/pylint$
 (master-fix-wrimposition-moy)$ flake8 --select=E402 pylint/test/functional/wrong_import_position*.py
pylint/test/functional/wrong_import_position11.py:4:1: E402 module level import not at top of file
pylint/test/functional/wrong_import_position12.py:5:1: E402 module level import not at top of file
pylint/test/functional/wrong_import_position13.py:4:1: E402 module level import not at top of file
pylint/test/functional/wrong_import_position14.py:5:1: E402 module level import not at top of file
pylint/test/functional/wrong_import_position.py:22:1: E402 module level import not at top of file
pylint/test/functional/wrong_import_position.py:26:1: E402 module level import not at top of file
pylint/test/functional/wrong_import_position.py:32:1: E402 module level import not at top of file
pylint/test/functional/wrong_import_position.py:33:1: E402 module level import not at top of file
[root@0cd245d5b6d4]~/build/PyCQA/pylint$
 (master-fix-wrimposition-moy)$ python -m pylint -d all -e wrong-import-position --output-format=colorized --msg-template="{pat
h}:{line}: [{msg_id}({symbol}), {obj}] {msg}" -r n pylint/test/functional/wrong_import_position*.py
************* Module functional.wrong_import_position11
pylint/test/functional/wrong_import_position11.py:4: [C0413(wrong-import-position), ] Import "import os" should be placed at the top of the module
************* Module functional.wrong_import_position12
pylint/test/functional/wrong_import_position12.py:5: [C0413(wrong-import-position), ] Import "import os" should be placed at the top of the module
************* Module functional.wrong_import_position13
pylint/test/functional/wrong_import_position13.py:4: [C0413(wrong-import-position), ] Import "from sys import x" should be placed at the top of the module
************* Module functional.wrong_import_position14
pylint/test/functional/wrong_import_position14.py:5: [C0413(wrong-import-position), ] Import "import y" should be placed at the top of the module
************* Module functional.wrong_import_position
pylint/test/functional/wrong_import_position.py:22: [C0413(wrong-import-position), ] Import "import six" should be placed at the top of the module
pylint/test/functional/wrong_import_position.py:26: [C0413(wrong-import-position), ] Import "import datetime" should be placed at the top of the module
pylint/test/functional/wrong_import_position.py:32: [C0413(wrong-import-position), ] Import "import scipy" should be placed at the top of the module
pylint/test/functional/wrong_import_position.py:33: [C0413(wrong-import-position), ] Import "import astroid" should be placed at the top of the module
```

screen shot 2016-05-17 at 8 28 19 pm

**Comment by moylop260** [comment]
@PCManticore 
I add all test cases from https://github.com/PyCQA/pylint/issues/901#issuecomment-219348588
I detected what we need `visit_expr` for the [wrong_import_position12.py](https://github.com/vauxoo-dev/pylint/blob/master-fix-wrimposition-moy/pylint/test/functional/wrong_import_position12.py) case.

Thanks for your feedback!

**Comment by moylop260** [comment]
Wait me a moment please, I will review the coverage decreased
coverage versus

**Comment by moylop260** [comment]
Well, Now I can sleep like a baby:

screen shot 2016-05-17 at 11 55 46 pm

**Comment by moylop260** [comment]
Anything else I should do?

**Comment by moylop260** [comment]
@PCManticore I'm ready to change this PR if is needed
Could you give me your feedback?

**Comment by PCManticore** [comment]
Looks okay, I just didn't have time to take a final look.
